File: leetcode/rb/20.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
  hash = {
    ")"=> "(",
    "}"=> "{",
    "]"=> "["
  }

  stack = []

  for ch in (s.split '')
    if hash.key? ch
      return false if stack.pop != hash[ch]
    else
      stack.push ch
    end
  end

  stack.size == 0
end

File: leetcode/py/20.py
class Solution:
    def isValid(self, s):
        stack = []

        pairs = {
            ')': '(',
            '}': '{',
            ']': '['
        }

        for ch in s:
            if ch not in pairs:
                stack.append(ch)
            elif pairs[ch] != stack.pop():
                return False

        return len(stack) == 0
